README.md
1. `java.awt.*`: The Abstract Window Toolkit (AWT) package provides classes for creating user interfaces and for painting graphics and images. It's a set of APIs to develop GUI or window-based applications in java.
2. `java.awt.event.ActionEvent`, `java.awt.event.ActionListener`, `java.awt.event.KeyEvent`, `java.awt.event.KeyListener`: These are part of the Java AWT event handling system. `ActionEvent` and `ActionListener` are used for handling button click events, while `KeyEvent` and `KeyListener` are used for handling keyboard events.
3. `java.io.FileWriter`: This class is used to write to character files. Its constructors can throw an `IOException` if the file to be written to cannot be found or some other I/O error occurs.
4. `java.io.IOException`: This is the general class of exceptions produced by failed or interrupted I/O operations.
5. `java.text.SimpleDateFormat`: This class provides methods to format and parse dates in a locale-sensitive manner.
6. `java.util.Date`: This class represents a specific instant in time, with millisecond precision.
7. `java.util.Random`: This class is used to generate a stream of pseudorandom numbers.
8. `javax.swing.*`: The Swing library is used for creating user interfaces. It provides a more sophisticated set of GUI components than the earlier AWT.
9. `java.io.BufferedReader`: This class reads text from a character-input stream, buffering characters so as to provide for the efficient reading of characters, arrays, and lines.
10. `java.io.FileReader`: This class is used to read character files. Its constructors can throw an `IOException` if the file to be read from cannot be found or some other I/O error occurs.


https://downloads.khinsider.com/
https://mixkit.co/
https://convertio.co/id/

changelog.md
# 🐍 2D Snake Game Changelog (Java Edition)

## 🚀 Version 2.0

### 🎉 New Features

🕰️ 09/12/2023, 19:40
- 📈 **Score History/Highest Score**: Keep track of your progress with our new score history and highest score feature.

🕰️ 09/12/2023, 20:07
- 🚀 **Snake Speed Levels**: Test your reflexes with our new snake speed levels.

🕰️ 09/12/2023, 20:30
- 🕹️ **Controller Functionality**: We've added controller support for a more immersive experience.

🕰️ 10/12/2023, 11:36
- 🖼️ **Revamped Game Banner**: The game now sports a sleek, new banner measuring 850px x 54px.

🕰️ 10/12/2023, 11:36
- 🎞️ **Detailed Animations**: We've added intricate animations to enhance the gaming experience.

🕰️ 10/12/2023, 11:57
- 🎮 **Game Start Change**: No more VK_RIGHT! Simply press space to start your game.

🕰️ 10/12/2023, 11:57
- 📝 **Player Record I/O**: Every player is now recorded thanks to our new I/O file processing.

🕰️ 10/12/2023, 13:15
- 🔄 **Exit or Re-Play Options**: Decide your next move with our new options to either exit or replay the game.

🕰️ 10/12/2023, 13:47
- 📊 **In-Game Level Indicator**: Keep track of your level with our new in-game level indicator.

🕰️ 10/12/2023, 14:55
- 📜 **Record Beautification**: We've given the record a facelift for easier reading.

🕰️ 10/12/2023, 16:34
- 🎨 **Game Icon**: Look out for our new game icon (1:1) on your desktop!

🕰️ 10/12/2023, 20:32
- 🐍 **Snake Body Reskin**: Our snake has undergone a makeover! Check out its new look.

🕰️ 11/12/2023, 00:50
- 🔊 **Sound Effects**: Immerse yourself in the game with our added sound effects.

🕰️ 11/12/2023, 02:03
- 🎵 **Soundtrack**: Enjoy our new, catchy soundtrack while you play.

🕰️ 11/12/2023, 15:07
- 🎲 **Game Modes**: Mix up your gameplay with our added game modes.

### 🐞 Bug Fixes

🕰️ 10/12/2023, 20:32
- 🖼️ **Gameplay Border**: We've fixed the gameplay border for a smoother gaming experience.

🕰️ 11/12/2023, 00:50
- 🐜 **Pixel Bug**: We've squashed the pesky 1 pixel bug.

### 🔄 Other Changes

🕰️ 10/12/2023, 11:36
- 📦 **Executable File**: We've exported the game into a .jar file and converted it into an executable for easier access.


main.java
import javax.swing.*;
import java.awt.*;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.UnsupportedAudioFileException;
import java.io.File;
import java.io.IOException;

public class main {
    public static void main(String[] args) {
        JFrame object = new JFrame();
        gamepy gp = new gamepy();

        object.setBounds(10,10,905,700);
        object.setBackground(Color.DARK_GRAY);
        object.setResizable(false);
        object.setVisible(true);
        object.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        object.add(gp);
    }

    public static void playSound(String soundFileName) {
        try {
            // Open an audio input stream.
            File soundFile = new File(soundFileName); //you could also get the sound file with an URL
            AudioInputStream audioIn = AudioSystem.getAudioInputStream(soundFile);              
            // Get a sound clip resource.
            Clip clip = AudioSystem.getClip();
            // Open audio clip and load samples from the audio input stream.
            clip.open(audioIn);
            clip.start();
        } catch (UnsupportedAudioFileException | IOException | LineUnavailableException e) {
            e.printStackTrace();
        }
    }
}

gamepy.java
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Random;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.UnsupportedAudioFileException;
import javax.swing.*;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;

class gamepy extends JPanel implements KeyListener, ActionListener {
    private int[] snakexlength = new int[750];
    private int[] snakeylength = new int[750];

    private int snakelength = 3;
    private boolean left = false;
    private boolean right = false;
    private boolean up = false;
    private boolean down = false;
    private int score = 0;
    private int level = 0;
    private int highestScore = 0;

    private int[] enemyxpos = {25,75,100,125,150,175,200,225,250,275,300,325,350,375,400,425,450,475,500,525,550,575,600,625,650,675,700,725,750};
    private int[] enemyypos = {75,100,125,150,175,200,225,250,275,300,325,350,375,400,425,450,475,500,525};

    private Random random = new Random();
    private int xpos = random.nextInt(29);
    private int ypos = random.nextInt(19);

    private Timer timer;
    private int delay = 100;
    private ImageIcon enemy;
    private ImageIcon rightimage;
    private ImageIcon upimage;
    private ImageIcon downimage;
    private ImageIcon leftimage;
    private ImageIcon snakeimage;
    private int move = 0;
    private ImageIcon topimage;

    private boolean gameStarted = false;
    private String playerName = "";
    private String lastRecordedPlayerName = "";
    private int lastRecordedScore = -1;
    private String lastRecordedDate = "";

    private Clip soundtrack;

    boolean isGameOver = false; // Add this line at the beginning of your class

    public gamepy() {
        addKeyListener(this);
        setFocusable(true);
        setFocusTraversalKeysEnabled(false);
        timer = new Timer(delay, this);
        timer.start();
    }

    public void paint(Graphics g) {
        if (!gameStarted) {
            g.setColor(Color.white);
            g.setFont(new Font("arial", Font.BOLD, 20));
            g.drawString("Press Space to Start!", 350, 300);
            return;
        }

        if (move == 0) {
            snakexlength[2] = 50;
            snakexlength[1] = 75;
            snakexlength[0] = 100;
            snakeylength[2] = 100;
            snakeylength[1] = 100;
            snakeylength[0] = 100;
        }

        // Draw title image border
        g.setColor(Color.white);
        g.drawRect(24, 10, 851, 55);

        // Draw the title image
        topimage = new ImageIcon("resource\\img\\snakebanner.png");
        topimage.paintIcon(this, g, 25, 11);

        // Draw border for gameplay
        g.setColor(Color.white);
        g.drawRect(24, 74, 851, 576);

        // Draw background for the gameplay
        g.setColor(Color.black);
        g.fillRect(25, 75, 850, 575);

        // Score record
        g.setColor(Color.white);
        g.setFont(new Font("arial", Font.PLAIN ,12));
        g.drawString(String.format("%-8s: %s", "Score", score), 700, 30);

        // Record of snake body's length
        g.setColor(Color.white);
        g.setFont(new Font("arial", Font.PLAIN ,12));
        g.drawString(String.format("%-7s: %s", "Length", snakelength), 700, 50);
        
        // Level indicator
        g.setColor(Color.white);
        g.setFont(new Font("arial", Font.PLAIN ,12));
        g.drawString(String.format("%-16s: %s", "Level", level), 778, 30);
        
        // Highest Score
        g.setColor(Color.white);
        g.setFont(new Font("arial", Font.PLAIN ,12));
        g.drawString(String.format("%-11s: %s", "Best Score", highestScore), 778, 50);

        // rightimage = new ImageIcon("head_right.png");
        // rightimage.paintIcon(this, g, snakexlength[0], snakeylength[0]);

        for (int a = 0; a < snakelength; a++) {
            if (a == 0) { // head of the snake
                if (right) {
                    rightimage = new ImageIcon("resource\\img\\graphics v3\\head_right.png");
                    rightimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                if (left) {
                    leftimage = new ImageIcon("resource\\img\\graphics v3\\head_left.png");
                    leftimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                if (up) {
                    upimage = new ImageIcon("resource\\img\\graphics v3\\head_up.png");
                    upimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                if (down) {
                    downimage = new ImageIcon("resource\\img\\graphics v3\\head_down.png");
                    downimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                } 
            } else if (a == snakelength - 1) { // tail of the snake
                if (right) {
                    snakeimage = new ImageIcon("resource\\img\\graphics v3\\tail_left.png");
                    snakeimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                if (left) {
                    snakeimage = new ImageIcon("resource\\img\\graphics v3\\tail_right.png");
                    snakeimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                if (up) {
                    snakeimage = new ImageIcon("resource\\img\\graphics v3\\tail_down.png");
                    snakeimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                if (down) {
                    snakeimage = new ImageIcon("resource\\img\\graphics v3\\tail_up.png");
                    snakeimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
            } else { // body of the snake
                if (right || left) {
                    snakeimage = new ImageIcon("resource\\img\\graphics v3\\body_horizontal.png");
                    snakeimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                if (up || down) {
                    snakeimage = new ImageIcon("resource\\img\\graphics v3\\body_vertical.png");
                    snakeimage.paintIcon(this, g, snakexlength[a], snakeylength[a]);
                }
                // Add conditions for bottomleft, bottomright, topleft, topright based on your game logic
            }
        }

        enemy = new ImageIcon("resource\\img\\apple.png");

        if (enemyxpos[xpos] == snakexlength[0] && enemyypos[ypos] == snakeylength[0]) {
            main.playSound("resource\\wav\\eat2.wav");
            score++;
            if (score > highestScore) {
                highestScore = score;
            }
            if (score % 10 == 0) {
                delay = (int)(delay * 0.9);
                level++;
                timer.setDelay(delay);
            }
            snakelength++;
            xpos = random.nextInt(29);
            ypos = random.nextInt(19);
        }

        enemy.paintIcon(this, g, enemyxpos[xpos], enemyypos[ypos]);

        for (int b = 1; b < snakelength; b++) {
            if (snakexlength[b] == snakexlength[0] && snakeylength[b] == snakeylength[0]) {
                right = left = up = down = false;
        
                if (!isGameOver) { // Add this line
                    main.playSound("resource\\wav\\gameover2.wav");
                    isGameOver = true; // And this line
                    soundtrack.stop();
                }
        
                // delay = 100;
                // timer.setDelay(delay);
        
                g.setColor(Color.white);
                g.setFont(new Font("arial", Font.BOLD, 50));
                g.drawString("GAME OVER!", 300, 300);
                g.setFont(new Font("arial", Font.BOLD, 20));
                g.drawString("Press SPACE to Re-Play", 350, 340);
                g.drawString("Press E to Exit", 350, 370);
        
                recordScore();
                repaint();
                level = 0;
            }
        }
        g.dispose();
    }
    public void playSoundtrack(String filename) {
        try {
            // If a soundtrack is already playing, stop it
            if (soundtrack != null && soundtrack.isRunning()) {
                soundtrack.stop();
                soundtrack.close();
            }
    
            // Start the new soundtrack
            AudioInputStream audioIn = AudioSystem.getAudioInputStream(new File(filename));
            soundtrack = AudioSystem.getClip();
            soundtrack.open(audioIn);
            soundtrack.loop(Clip.LOOP_CONTINUOUSLY);
        } catch (UnsupportedAudioFileException | IOException | LineUnavailableException e) {
            e.printStackTrace();
        }
    }

    public void setGameMode() {
        String[] gameModes = {"Easy Mode", "Intermediate", "Hard Mode"};
        int selectedMode = JOptionPane.showOptionDialog(null, "Choose a game mode", "Game Mode",
                JOptionPane.DEFAULT_OPTION, JOptionPane.INFORMATION_MESSAGE, null, gameModes, gameModes[0]);
        switch (selectedMode) {
            case 0: // Easy Mode
                delay *= 1; // No speed change
                break;
            case 1: // Intermediate
                delay *= 0.8; // Increase speed by 20%
                break;
            case 2: // Hard Mode
                delay *= 0.6; // Increase speed by 40%
                break;
        }
        timer.setDelay(delay);
    }

    private int recordNo = 0;

    private void recordScore() {
        try {
            // Read the current record number from the scores.txt file
            BufferedReader reader = new BufferedReader(new FileReader("scores.txt"));
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split("\\s+");
                if (parts.length > 0) {
                    String recordNoStr = parts[0];
                    if (recordNoStr.matches("\\d+")) {
                        int currentRecordNo = Integer.parseInt(recordNoStr);
                        if (currentRecordNo > recordNo) {
                            recordNo = currentRecordNo;
                        }
                    }
                }
            }
            reader.close();

            if (!playerName.equals(lastRecordedPlayerName) || score != lastRecordedScore || !getCurrentDate().equals(lastRecordedDate)) {
                FileWriter writer = null;
                try {
                    writer = new FileWriter("scores.txt", true);
                    String currentTime = getCurrentTime();
                    String record = String.format("%-4d %-20s %-8d %-8d %s %-4s\n", ++recordNo, playerName, score, level, getCurrentDate(), currentTime);
                    writer.write(record);
                    writer.close();
                    lastRecordedPlayerName = playerName;
                    lastRecordedScore = score;
                    lastRecordedDate = getCurrentDate();
                } catch (IOException e) {
                    e.printStackTrace();
                } finally {
                    if (writer != null) {
                        try {
                            writer.close();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private String getCurrentTime() {
        return new SimpleDateFormat("HH:mm:ss").format(new Date());
    }

    private String getCurrentDate() {
        return new SimpleDateFormat("yyyy-MM-dd").format(new Date());
    }

    @Override
    public void actionPerformed(ActionEvent arg0) {
        timer.start();
        if (right) {
            for (int b = snakelength - 1; b >= 0; b--) {
                snakeylength[b + 1] = snakeylength[b];
            }
            for (int b = snakelength; b >= 0; b--) {
                if (b == 0) {
                    snakexlength[b] = snakexlength[b] + 25;
                } else {
                    snakexlength[b] = snakexlength[b - 1];
                }
                if (snakexlength[b] > 850) {
                    snakexlength[b] = 25;
                }
            }
            repaint();
        }
        if (left) {
            for (int b = snakelength - 1; b >= 0; b--) {
                snakeylength[b + 1] = snakeylength[b];
            }
            for (int b = snakelength; b >= 0; b--) {
                if (b == 0) {
                    snakexlength[b] = snakexlength[b] - 25;
                } else {
                    snakexlength[b] = snakexlength[b - 1];
                }
                if (snakexlength[b] < 25) {
                    snakexlength[b] = 850;
                }
            }
            repaint();
        }
        if (up) {
            for (int b = snakelength - 1; b >= 0; b--) {
                snakexlength[b + 1] = snakexlength[b];
            }
            for (int b = snakelength; b >= 0; b--) {
                if (b == 0) {
                    snakeylength[b] = snakeylength[b] - 25;
                } else {
                    snakeylength[b] = snakeylength[b - 1];
                }
                if (snakeylength[b] < 75) {
                    snakeylength[b] = 625;
                }
            }
            repaint();
        }
        if (down) {
            for (int b = snakelength - 1; b >= 0; b--) {
                snakexlength[b + 1] = snakexlength[b];
            }
            for (int b = snakelength; b >= 0; b--) {
                if (b == 0) {
                    snakeylength[b] = snakeylength[b] + 25;
                } else {
                    snakeylength[b] = snakeylength[b - 1];
                }
                if (snakeylength[b] > 625) {
                    snakeylength[b] = 75;
                }
            }
            repaint();
        }
    }

    @Override
    public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_SPACE) {
            if (!gameStarted) {
                playerName = JOptionPane.showInputDialog("Enter your name:");
                setGameMode();
                playSoundtrack("resource\\wav\\soundtrack1.wav");
                gameStarted = true;
                // Hide the dialog box
                JOptionPane.getRootFrame().dispose();
                repaint();
            } else {
                move = 0;
                score = 0;
                snakelength = 3;
                isGameOver = false;
                playSoundtrack("resource\\wav\\soundtrack1.wav");
            }
        } else if (e.getKeyCode() == KeyEvent.VK_E) {
            System.exit(0); // Exit the game
        } else {
            if (gameStarted) {
                if (e.getKeyCode() == KeyEvent.VK_RIGHT) {
                    move++;
                    right = true;
                    if (!left) {
                        right = true;
                    } else {
                        right = false;
                        left = true;
                    }
                    up = false;
                    down = false;
                } else if (e.getKeyCode() == KeyEvent.VK_LEFT) {
                    move++;
                    left = true;
                    if (!right) {
                        left = true;
                    } else {
                        left = false;
                        right = true;
                    }
                    up = false;
                    down = false;
                } else if (e.getKeyCode() == KeyEvent.VK_UP) {
                    move++;
                    up = true;
                    if (!down) {
                        up = true;
                    } else {
                        up = false;
                        down = true;
                    }
                    right = false;
                    left = false;
                } else if (e.getKeyCode() == KeyEvent.VK_DOWN) {
                    move++;
                    down = true;
                    if (!up) {
                        down = true;
                    } else {
                        down = false;
                        up = true;
                    }
                    right = false;
                    left = false;
                }
            }
        }
    }

    @Override
    public void keyReleased(KeyEvent arg0) {
    }

    @Override
    public void keyTyped(KeyEvent arg0) {
    }
}